packages/lodestar/src/chain/genesis/genesis.ts

```diff
@@ -28,16 +28,167 @@ import {
+
+
+  /**
+   * This processes an eth1 block and check if it can form genesis state of eth1.
+   */
+  public isFormingValidGenesisState = async (events: IDepositEvent[]): Promise<boolean> => {
+    const block = await this.eth1.getBlock(events[0].blockNumber);
+    applyTimestamp(this.config, this.state, block.timestamp);
+    applyEth1BlockHash(this.config, this.state, fromHexString(block.hash));
+    const isValid = isValidGenesisState(this.config, this.state);
+    this.logger.verbose(`genesis: Process block ${block.number}, valid genesis state=${isValid}`);
+    if (isValid) {
+      const eth1Data = {
+        blockHash: fromHexString(block.hash),
+        depositRoot: this.depositTree.tree().root,
+        depositCount: events[events.length - 1].index + 1,
+      };
+
+      await Promise.all([
+        // proposing data
+        this.db.depositData.batchPut(events.map((depositEvent) => ({
+          key: depositEvent.index,
+          value: depositEvent,
+        }))),
+        this.db.eth1Data.put(block.timestamp, eth1Data)
+      ]);
+    }
+    return isValid;
+  };
+
+  public processDepositEvents(): ((source: AsyncIterable<IDepositEvent[]>) => AsyncIterable<IDepositEvent[]>) {
+    const {logger, doProcessDepositEvents, config} = this;
+    return ((source: AsyncIterable<IDepositEvent[]>): AsyncIterable<IDepositEvent[]> => {
+      return async function* () {
+        for await (const depositEvents of source) {
+          const lastBlockNumber = depositEvents[depositEvents.length - 1].blockNumber;
+          const firstBlockNumber = depositEvents[0].blockNumber;
+          for (let blockNumber = firstBlockNumber; blockNumber <= lastBlockNumber; blockNumber++) {
+            const blockDepositEvents = depositEvents.filter(event => event.blockNumber === blockNumber);
+            if (blockDepositEvents.length > 0) {
+              await doProcessDepositEvents(blockDepositEvents);
+              const depositCount = blockDepositEvents[blockDepositEvents.length - 1].index + 1;
+              logger.info(`genesis: Found ${blockDepositEvents.length} deposit events for block ${blockNumber}, ` +
+                `depositCount=${depositCount}`);
+              if (depositCount >= config.params.MIN_GENESIS_ACTIVE_VALIDATOR_COUNT) {
+                yield blockDepositEvents;
+              }
+            }
+          }
+        }
+      }();
+    });
+  }
+
+  public genesis = async (): Promise<TreeBacked<BeaconState>> => {
+    await this.initialize();
+    const eth1DataStream = await this.eth1.getDepositEventsFromBlock(true, undefined);
+    const {isFormingValidGenesisState: processSingleBlock, state} = this;
+    const foundGenesis = this.eth1.foundGenesis.bind(this.eth1);
+    return await pipe(eth1DataStream,
+      this.processDepositEvents(),
+      async function(source: AsyncIterable<IDepositEvent[]>) {
+        for await (const depositEvents of source) {
+          const isValidGenesis = await processSingleBlock(depositEvents);
```

Comment:
this line makes no sense processSingleBlock that receives deposit events returns isValidGenesis

Reply:
this is to process deposit events of a single block, and we accumulate deposits to the cached state and check if that state is genesis.
I guess the confusion comes from the name. How do you want to name it?